numpy/_core/src/multiarray/einsum_sumprod.cpp

```diff
@@ -0,0 +1,2181 @@
+/*
+ * This file provides optimized sum of product implementations used internally
+ * by einsum.
+ *
+ * Copyright (c) 2011 by Mark Wiebe (mwwiebe@gmail.com)
+ * The University of British Columbia
+ *
+ * See LICENSE.txt for the license.
+ *
+ */
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#include "einsum_sumprod.h"
+
+#include "common.h"
+#include "einsum_debug.h"
+
+#include "simd/simd.h"
+#include <array>
+#include <iostream>
+#include <numpy/arrayobject.h>
+#include <numpy/halffloat.h>
+#include <numpy/ndarraytypes.h> /* for NPY_NTYPES_LEGACY */
+#include <numpy/npy_common.h>
+
+#include "simd_accum_traits.h"
+
+// ARM/Neon don't have instructions for aligned memory access
+#ifdef NPY_HAVE_NEON
+#    define EINSUM_IS_ALIGNED(x) 0
+#else  // NPY_HAVE_NEON
+#    define EINSUM_IS_ALIGNED(x) npy_is_aligned(x, NPY_SIMD_WIDTH)
+#endif  // NPY_HAVE_NEON
+
+/* Converts a value from storage type T to the AccumType.
+ * Centralizes norminal use and the specalized use case of npy_half_to_float
+ */
+template <typename T, typename AccumType>
+static inline constexpr AccumType
+from(T v)
+{
+    return (AccumType)v;
+}
+
+template <>
+inline npy_float
+from<npy_half, npy_float>(npy_half v)
+{
+    return npy_half_to_float(v);
+}
+
+/* Converts a value from storage type AccumType to the Type.
+ * Centralizes norminal use and the specalized use case of npy_float_to_half
+ */
+template <typename T, typename AccumType>
+static inline constexpr AccumType
+to(T v)
+{
+    return (AccumType)v;
+}
+
+template <>
+inline npy_half
+to<npy_float, npy_half>(npy_float v)
+{
+    return npy_float_to_half(v);
+}
+
+template <typename SimdType>
+static inline NPY_GCC_OPT_3 typename SimdType::T
+floating_point_sum_of_arr(const typename SimdType::T *data, npy_intp count)
```

Comment:
```suggestion
template <typename Lane>
static inline NPY_GCC_OPT_3 Lane
floating_point_sum_of_arr(const Lane *data, ptrdiff_t count)
```
Just deduce the scalar type directly; regarding `npy_intp`, you can use `SSize`, which maps to `Py_ssize_t` or `ptrdiff_t`.